File: app/src/main/java/com/sample/app/fragment_with_retrofit/FragmentWithRetrofit.java
package com.sample.app.fragment_with_retrofit;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.sample.app.R;
import com.sample.app.api.RetrofitUtil;
import com.sample.app.api.customer.CustomerAPI;
import com.sample.app.db.repository.CustomerRepository;
import com.sample.app.entity.Customer;
import com.sample.app.fragment_with_retrofit.runnables.SaveCustomerRunnable;
import com.sample.app.utils.GsonUtils;
import com.google.gson.JsonElement;
import com.google.gson.reflect.TypeToken;

import java.util.List;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import retrofit2.Retrofit;

public class FragmentWithRetrofit extends Fragment {

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        Retrofit retrofit = RetrofitUtil.getSimpleRetrofit();
        CustomerAPI customerAPI = retrofit.create(CustomerAPI.class);

        //Call<JsonElement> call = customerAPI.getCustomers1();
        Call<List<Customer>> call = customerAPI.getCustomers2();
        call.enqueue(custCallback);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {

        View rootView      = inflater.inflate(R.layout.fragment_with_retrofit, container, false);
        return rootView;
    }

    Callback<JsonElement> jsonCallback = new Callback<JsonElement>() {
        @Override
        public void onResponse(Call<JsonElement> call, Response<JsonElement> response) {
            Log.d("response","onResponse");
            if (response.body() != null) {
                JsonElement json = response.body();
                Log.d("response","response :" + json);
                List<Customer> customerList = GsonUtils.getGson().fromJson(json, new TypeToken<List<Customer>>(){}.getType());
                saveCustomerToDb(customerList);
            }
        }

        @Override
        public void onFailure(Call<JsonElement> call, Throwable t) {
            Log.d("response","onFailure");
            getCustomerfromDB();
        }
    };

    Callback<List<Customer>> custCallback = new Callback<List<Customer>>() {
        @Override
        public void onResponse(Call<List<Customer>> call, Response<List<Customer>> response) {
            Log.d("response","onResponse");
            if (response.body() != null) {
                List<Customer> customerList = response.body();
                saveCustomerToDb(customerList);
            }
        }

        @Override
        public void onFailure(Call<List<Customer>> call, Throwable t) {
            Log.d("response","onFailure");
            getCustomerfromDB();
        }
    };

    private void saveCustomerToDb(List<Customer> customerList) {
        SaveCustomerRunnable saveCustomerRunnable = new SaveCustomerRunnable(getContext(), customerList);
        Thread thread = new Thread(saveCustomerRunnable);
        thread.start();
    }

    private void getCustomerfromDB() {
        List<Customer> customerList = CustomerRepository.getAll(getContext());
    }
}

File: app/src/main/java/com/sample/app/fragment_with_retrofit/FragmentWithRetrofit.kt
package com.sample.app.fragment_with_retrofit

import android.os.Bundle
import android.support.v4.app.Fragment
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup

import com.google.gson.JsonElement
import com.google.gson.reflect.TypeToken
import com.sample.app.R
import com.sample.app.api.RetrofitUtil
import com.sample.app.api.customer.CustomerAPI
import com.sample.app.entity.Customer
import com.sample.app.fragment_with_retrofit.runnables.SaveCustomerRunnable
import com.sample.app.repository.CustomerRepository
import com.sample.app.utils.GsonUtils

import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class FragmentWithRetrofit : Fragment() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        val retrofit = RetrofitUtil.simpleRetrofit
        val customerAPI = retrofit.create(CustomerAPI::class.java)

        val call = customerAPI.customers1
        //Call<List<Customer>> call = customerAPI.getCustomers2();
        call.enqueue(jsonCallback)
    }

    override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {

        val rootView = inflater!!.inflate(R.layout.fragment_with_retrofit, container, false)
        return rootView
    }

    internal var jsonCallback: Callback<JsonElement> = object : Callback<JsonElement> {
        override fun onResponse(call: Call<JsonElement>, response: Response<JsonElement>) {
            Log.d("response", "onResponse")
            if (response.body() != null) {
                val json = response.body()
                Log.d("response", "response :" + json)
                val customerList = GsonUtils.getGson()!!.fromJson<List<Customer>>(json, object : TypeToken<List<Customer>>() {

                }.type)
                saveCustomerToDb(customerList)
            }
        }

        override fun onFailure(call: Call<JsonElement>, t: Throwable) {
            Log.d("response", "onFailure")
            getCustomerfromDB()
        }
    }

    internal var custCallback: Callback<List<Customer>> = object : Callback<List<Customer>> {
        override fun onResponse(call: Call<List<Customer>>, response: Response<List<Customer>>) {
            Log.d("response", "onResponse")
            if (response.body() != null) {
                val customerList = response.body()
                if (customerList != null) {
                    saveCustomerToDb(customerList)
                }
            }
        }

        override fun onFailure(call: Call<List<Customer>>, t: Throwable) {
            Log.d("response", "onFailure")
            getCustomerfromDB()
        }
    }

    private fun saveCustomerToDb(customerList: List<Customer>) {
        val saveCustomerRunnable = SaveCustomerRunnable(context, customerList)
        val thread = Thread(saveCustomerRunnable)
        thread.start()
    }

    private fun getCustomerfromDB() {
        val customerList = CustomerRepository.all
    }
}